Subscribe to streaming updates by routeId and stopId

thecristen opened the pull request:
<!-- 
  GitHub will automatically add a random non-busy member of the Dotcom team
  as a Required Reviewer when the PR is opened; feel free to also manually assign
  team members if a PR needs a very specific pair of eyes! -->

#### Summary of changes

<!-- Link to relevant Asana task; remove if not applicable -->
**Asana Ticket:** [Subscribe to streaming updates by routeId and stopId](https://app.asana.com/0/385363666817452/1203082255397801/f)


> **Warning**
> This was substantially reworked Dec 14! Description has been updated below.

This builds upon the work done in #1421. To that end, the first three commits are directly from Matt's branch there. The fourth commit (`fixup: feat: Parse and save prediction arrival and departure times`) conceptually goes with the changes in the third commit. 

So, the fifth commit (`feat: get streaming predictions by route, stop, direction`) contains the bulk of the task to enable subscribing to a stream of predictions for a specified stop on a route, rather than all predictions for an entire route. This was further enhanced to add a direction argument. This mostly involved adding a stop and direction arguments to wherever we were using only route ID.

The code uses a channel name format of `"predictions:routeId:stopId:directionId"` e.g. `predictions:Red:place-sstat:1`.

The last two commits are additional enhancements that I feel complete the implementation. These:
* Ensure we use the _parent_ stop when we're parsing the streamed predictions.
* Push the predictions to the socket using the "data" event name, just for consistency with our other channel.

Note this is a backend-only implementation. I'll post another PR soon for the frontend piece of this, which will also build upon the frontend pieces of #1421 but will be a bit more significant of a change.

---

#### General checks
* [x] **Are the changes organized into self-contained commits with descriptive and well-formatted commit messages?** This is a good practice that can facilitate easier reviews.
* [x] **Testing.** Do the changes include relevant passing updates to tests? This includes updating screenshots. Preferably tests are run locally to verify that there are no test failures created by these changes, before opening a PR.
* [x] **Tech debt.** Have you checked for tech debt you can address in the area you're working in? This can be a good time to address small issues, or create Asana tickets for larger issues.

i0sea reviewed (commented):
Does [this](https://github.com/mbta/dotcom/pull/1421) need to be merged first? See some overlap in changes here.

thecristen commented on apps/predictions/lib/parser.ex:
```diff
@@ -13,30 +14,83 @@ defmodule Predictions.Parser do
           DateTime.t() | nil,
           non_neg_integer,
-          Prediction.schedule_relationship(),
+          Prediction.schedule_relationship() | nil,
```
Not sure, but `Prediction.schedule_relationship()` includes the `nil` type anyway so looks like this change isn't needed. 🤔

arkadyan replied:
@thecristen Yep, I must have missed that. Agreed that this change isn't needed.

thecristen commented:
Inviting @arkadyan to review in case he has any thoughts or context. Thanks for making the foundation of this work we're finally getting around to revisiting! 😉

thecristen commented:
@i0sea I've added some more detail to the description and updated everything. Basically this would replace part of the #1421, so we don't expect to merge that in first.

arkadyan commented on apps/predictions/lib/stream_parser.ex:
```diff
@@ -68,10 +68,18 @@ defmodule Predictions.StreamParser do
 
   defp trip(_), do: nil
 
+  # Predictions are generally associated in the API with child stops. Find the
```
👍

arkadyan commented on apps/predictions/lib/predictions_pub_sub.ex:
```diff
@@ -32,12 +36,15 @@ defmodule Predictions.PredictionsPubSub do
+    key = prediction_key(route_id, stop_id, direction_id)
+    StreamSupervisor.ensure_stream_is_started(key)
+    {registry_key, predictions} = GenServer.call(server, {:subscribe, key})
+    Registry.register(:prediction_subscriptions_registry, registry_key, key)
```
suggestion (non-blocking): If you wanted to, you could store predictions by route+stop+direction, while leaving the streaming as by route. In other words, the server-client contract could be based on route+stop+direction, while the server-API client would just be based on route. The con here would be that you'd be streaming more data than you need from the API, but the pro would be less churn in starting small streams. I'm not sure what the performance implications would be in practice, but my gut instinct is that there are very likely to be multiple users looking at the same route, even if they are looking at different stops along it.

thecristen replied:
@arkadyan I ended up doing this!

arkadyan commented on apps/predictions/lib/predictions_pub_sub.ex:
```diff
@@ -1,20 +1,24 @@
 
   @type t :: %__MODULE__{
-          predictions_by_route_id: predictions_by_route_id()
+          predictions_by_key: predictions_by_key()
```
suggestion (non-blocking): At some point (especially if we see performance issues with this), it might be worth considering using an `ets` table for this data. It could use `route`, `stop`, and `direction` as keys.

thecristen replied:
We'd have to read/write to the table super frequently no? An interesting thing to keep in mind for sure! Let's see how well this works out without ETS first.

paulswartz replied:
reading/writing super-frequently (especially from different processes) is basically the reason you'd use ETS :)

arkadyan reviewed (commented):
I did a little more skimming than going totally in-depth, but this looks pretty great to me. I threw out a couple of comments you can consider and take or leave, whether now or in the future.

thecristen commented on apps/predictions/lib/stream_supervisor.ex:
```diff
@@ -0,0 +1,100 @@
+      "filter[route]=#{route_id}&filter[stop]=#{stop_id}&filter[direction_id]=#{direction_id}"
+
+    path =
+      "/predictions?#{filters}&fields[prediction]=status,departure_time,arrival_time,direction_id,schedule_relationship,stop_sequence&include=route,trip,trip.occupancies,stop&fields[route]=long_name,short_name,type&fields[trip]=direction_id,headsign,name,bikes_allowed&fields[stop]=platform_code"
```
I was reading the [documentation](https://api-v3.mbta.com/docs/swagger/index.html#/Prediction/ApiWeb_PredictionController_index) and realized the API already offers the `time` field where it uses `arrival_time` if available otherwise using `departure_time` (aka implementing our logic for "first" time). Maybe we should use that.

arkadyan commented:
Asking Eddie to review instead of me for a fresh pair of eyes since I was involved early on.

lemald commented on apps/predictions/lib/parser.ex:
```diff
@@ -13,30 +14,83 @@ defmodule Predictions.Parser do
+  def departing?(%Item{attributes: %{"departure_time" => binary}}) when is_binary(binary),
+    do: true
+
+  def departing?(%Item{attributes: %{"status" => binary}}) when is_binary(binary),
+    do: upcoming_status?(binary)
```
**question:** could you speak a bit to how the logic in `upcoming_status?` related to whether or not a prediction corresponds to a departure? It looks like it might have something to do with Commuter Rail, but then `upcoming_status?` also checks for the string "away" in the status which looks to me like it might be related to "stops away" on the rapid transit?

thecristen replied:
That was added six years ago in https://github.com/mbta/dotcom-legacy/pull/1004/ :D 
It look like our current code used it as a way to guess whether a prediction corresponds to a departure, for predictions which have no departure time.
Do you suppose it might no longer be relevant?

lemald replied:
Ahh okay, I think that the "away" part at least must date from a different era when we didn't yet have minute countdowns on the Green Line but instead displayed [only the number of stops away](https://upload.wikimedia.org/wikipedia/commons/1/1f/Stops_away_countdown_signs_at_Park_Street%2C_January_2016.JPG). So that is likely no longer relevant. And that sign also shows "Approaching" and "Boarding."

lemald replied:
So I think from a subway perspective we can probably ditch this logic. I would just want to make sure there are no unintended consequences with Commuter Rail data since we also use the `status` field there.

lemald replied:
Also this PR has already got a lot of stuff going on and this logic isn't actually being changed as part of it, so I think it would be fine if you punted this to a tech debt task.

lemald commented on apps/predictions/lib/stream_supervisor.ex:
```diff
@@ -0,0 +1,84 @@
+  def ensure_stream_is_started(route_id),
+    do: ensure_stream_is_started(route_id, System.get_env("USE_SERVER_SENT_EVENTS"))
+
+  defp ensure_stream_is_started(_route_id, "false"), do: :bypassed
```
**question:** do I understand correctly that this means that the application will use SSEs unless the `USE_SERVER_SENT_EVENTS` environment variable is explicitly set to `"false"`?

thecristen replied:
Yes, exactly. We happen to use a similar pattern for vehicle data.

lemald replied:
Okay cool, I was sort of confused by it but as long as it's following an established pattern that's fine.

lemald commented on apps/predictions/lib/stream_supervisor.ex:
```diff
@@ -0,0 +1,84 @@
+
+  @spec ensure_stream_is_started(Route.id_t()) :: {:ok, pid()} | :bypassed
+  def ensure_stream_is_started(route_id),
+    do: ensure_stream_is_started(route_id, System.get_env("USE_SERVER_SENT_EVENTS"))
```
**suggestion (non-blocking):** I'm not sure exactly how dotcom is built and deployed and how releases factor into it (or don't), but if you want to be more Elixir-y about this it may be possible to [set the application configuration at runtime](https://hexdocs.pm/mix/1.14/Mix.Tasks.Release.html#module-runtime-configuration) based on the environment variable rather than querying it directly here.

thecristen replied:
Yes!! Not ready to tackle this yet.

lemald commented on apps/predictions/lib/stream_supervisor.ex:
```diff
@@ -0,0 +1,84 @@
+  @spec sses_opts(Route.id_t()) :: Keyword.t()
+  defp sses_opts(route_id) do
+    path =
+      "/predictions?filter[route]=#{route_id}&fields[prediction]=status,departure_time,arrival_time,direction_id,schedule_relationship,stop_sequence&include=route,trip,trip.occupancies,stop&fields[route]=direction_destinations,direction_names,long_name,short_name,type&fields[trip]=direction_id,headsign,name,bikes_allowed&fields[stop]=platform_code"
```
**issue:** you probably want to use something like `URI.encode_query/2` here, I don't think any of our route IDs contain problematic characters in practice but per [the GTFS spec](https://gtfs.org/schedule/reference/#field-types) IDs can be arbitrary strings.

thecristen replied:
Oh interesting! 
I've added a commit to use `URI.encode/1` on the URL, but this change only affects URLs we send to `ServerSentEventStage`.

(Our other requests are part of a different workflow that would've been trickier to adjust. But that's using `HTTPoison.Base` which looks like it might have its own mechanism for at least checking the URL is parseable. I suppose it's yet another thing we can handle later!)

lemald commented on apps/predictions/test/predictions_pub_sub_test.exs:
```diff
@@ -0,0 +1,168 @@
+    end
+
+    test "updates the predictions", %{pid: pid} do
+      send(pid, {:update, [@prediction39]})
```
**question:** am I reading this correctly that the initial state is `@prediction39` and it's being replaced with `@prediction39`? The test should probably take a modified prediction value to confirm that the update actually works.

thecristen replied:
💯 hah, great catch! updated.

lemald commented on apps/site/test/support/env_helpers.ex:
```diff
@@ -0,0 +1,18 @@
+defmodule Test.Support.EnvHelpers do
+  @moduledoc "Helpers for reassigning env variables"
+
+  defmacro reassign_env(app, var, value) do
```
**praise:** love having this macro handy for writing tests

lemald commented on apps/predictions/lib/prediction.ex:
```diff
@@ -25,6 +27,9 @@ defmodule Predictions.Prediction do
           direction_id: 0 | 1,
+          arrival_time: DateTime.t() | nil,
+          departure_time: DateTime.t() | nil,
+          # TODO: Deprecated, should be removed in favor of arrival_time and departure_time -- MSS 2022-09-22
```
**suggestion:** I'm not sure how your team tracks this sort of cleanup work, but if this field should be removed eventually is there a ticket to represent that?

thecristen replied:
There is now! :D

lemald commented on apps/predictions/lib/predictions_pub_sub.ex:
```diff
@@ -1,20 +1,24 @@
   @moduledoc """
-  Allow channels to subscribe to prediction streams.
+  Allow channels to subscribe to prediction streams, which are collected into a
+  Map with keys representing route-stop pairs as "<route-id>@<parent-stop-id>"
```
**nitpick:** this should be updated to reflect the actual implementation of `prediction_key/3`.

lemald commented on apps/site/lib/site_web/channels/predictions_channel.ex:
```diff
@@ -21,7 +21,11 @@ defmodule SiteWeb.PredictionsChannel do
+    future_predictions = Enum.filter(predictions, &is_in_future?/1)
+    # immediately push current predictions to socket
+    send(self(), {:new_predictions, future_predictions})
+    {:ok, socket}
```
**question:** I'm a bit confused about this change. Returning `{:ok, %{predictions: predictions}, socket}` should already make the data available to the client as soon as it joins, though it's a little bit confusing because it's accessed in a different way from the usual results of a push. We actually had a bug in Skate caused by this issue but we were able to fix it client-side in mbta/skate#1575.

thecristen replied:
That's from an older commit -- the current code removes the `future_predictions = Enum.filter(predictions, &is_in_future?/1)` from here and puts it in the `handle_info({:new_predictions, predictions}, socket)` clause, which is also handling some other data transformation. To clarify, we `send(self(), {:new_predictions, predictions})` so that the handle_info can do the data filtering before sending the data out. I hope this helps?

thecristen replied:
Ahhhh, thanks for elaborating. That makes sense now. Have added a commit to make this change, great suggestion 👍🏼

lemald commented:
@thecristen @arkadyan okay, it took a while but I think I'm through my initial pass of reviewing!

lemald commented on apps/predictions/lib/predictions_pub_sub.ex:
```diff
@@ -0,0 +1,239 @@
+  @moduledoc """
+  Allow channels to subscribe to prediction streams, which are collected into a
+  Map with keys representing route-stop-direction triads as
+  "<route-id>-<parent-stop-id>-<direction-id>"
```
**super-duper nitpick (non-blocking):** this now matches the format from `prediction_key/3`, but the actual implementation uses colons rather than hyphens.

Also just as a thought, maybe you could just use tuples instead of strings as the map keys? Unless there's some place they need to be serialized as strings, but maybe even them you could do that as a separate step.
